core/blockchain.go

```diff
@@ -290,17 +400,30 @@ func (self *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
 	return nil
 }
 
-// insert injects a block into the current chain block chain. Note, this function
-// assumes that the `mu` mutex is held!
+// insert injects a new head block into the current block chain. This method
+// assumes that the block is indeed a true head. It will also reset the head
+// header and the head fast sync block to this very same block to prevent them
+// from diverging on a different header chain.
```

Comment:
Changed to

> It will also reset the head header and the head fast sync block to this very same block to prevent them from pointing to a possible old canonical chain (i.e. side chain by now).